### PR TITLE
Support vicuna dataset_format

### LINKED ISSUE #104: DataCollator inappropriate for Chat Data?
Hello, great paper and repository, thanks. 

I see that the data collator seems to be targeting the instruction tuning use-case where we have an `input` explaining the task and a single `output` solving it. By default we reasonably train only on that output using `IGNORE_INDEX` on the tokens of the `input` part.

However this breaks for chat kind of datasets and since one of the paper's good results is related to OASST1 I wonder whether strong improvements can trivially be achieved by editing the collator. Right now the whole (potentially multi-round) conversation is dumped into the `output` field that the model trains on, nothing is masked by `IGNORE_INDEX`. So training treats the role asking question (user) the same as the role answering (assistant) by training on both. Is that desirable? Why not stick with Vicuna chat-oriented `preprocess` function like [here](https://github.com/lm-sys/FastChat/blob/main/fastchat/train/train.py#L77) which only conditions on the role asking but doesn't back-propagate loss computed oon it?

Apologies if my understanding is incorrect or limited.

### PR DESCRIPTION
**Pull request by jieguangzhou**
I add a new dataset format function to handle the Vicuna dataset

Because I want to use Qlora to train vicuna13B and vicuna7B, but if I use instruction data, I think that would lose some Conversational ability.

close: #104

### CODE REVIEW

**Comment by zhangqibupt** [comment]
Thank you for the significant improvements you've made. I appreciate your efforts. 
I am currently in the process of testing your branch on my local environment. Although the model can be loaded successfully, see line 
`loading base model ../FastChat/models/vicuna-7b-all-v1.1...`

However I'm facing an issue with loading the dataset. It seems that the script is not loading the file I specified. See
`Found cached dataset json (/home/ubuntu/.cache/huggingface/datasets/json/default-90b8cf6b2bdf123d/0.0.0)
Loading cached split indices for dataset at /home/ubuntu/.cache/huggingface/datasets/json/default-`

Unfortunately, I'm unable to determine the cause of this problem. Could you kindly provide assistance in resolving this matter? Your help would be greatly appreciated. Thank you in advance.

```
python qlora.py \
>   --model_name_or_path='../FastChat/models/vicuna-7b-all-v1.1' \
>   --dataset='../FastChat/fastchat/train_data/train.json' \
>   --dataset_format='vicuna'


===================================BUG REPORT===================================
Welcome to bitsandbytes. For bug reports, please run

python -m bitsandbytes

 and submit this information together with your error trace to: https://github.com/TimDettmers/bitsandbytes/issues
================================================================================
bin /opt/conda/lib/python3.10/site-packages/bitsandbytes/libbitsandbytes_cuda117.so
CUDA SETUP: CUDA runtime path found: /opt/conda/lib/libcudart.so.11.0
CUDA SETUP: Highest compute capability among GPUs detected: 7.5
CUDA SETUP: Detected CUDA version 117
CUDA SETUP: Loading binary /opt/conda/lib/python3.10/site-packages/bitsandbytes/libbitsandbytes_cuda117.so...
loading base model ../FastChat/models/vicuna-7b-all-v1.1...
The model weights are not tied. Please use the `tie_weights` method before using the `infer_auto_device` function.
Loading checkpoint shards: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:19<00:00,  9.84s/it]
adding LoRA modules...
trainable params: 79953920.0 || all params: 3660320768 || trainable: 2.184341894267557
loaded model
Adding special tokens.
Found cached dataset json (/home/ubuntu/.cache/huggingface/datasets/json/default-90b8cf6b2bdf123d/0.0.0)
Loading cached split indices for dataset at /home/ubuntu/.cache/huggingface/datasets/json/default-90b8cf6b2bdf123d/0.0.0/cache-dbe043eb54573d57.arrow and /home/ubuntu/.cache/huggingface/datasets/json/default-90b8cf6b2bdf123d/0.0.0/cache-3a99fdef2139a825.arrow
Loading cached processed dataset at /home/ubuntu/.cache/huggingface/datasets/json/default-90b8cf6b2bdf123d/0.0.0/cache-7f95e7149fd12742.arrow
Loading cached processed dataset at /home/ubuntu/.cache/huggingface/datasets/json/default-90b8cf6b2bdf123d/0.0.0/cache-86ed03a862573a38.arrow
Loading cached processed dataset at /home/ubuntu/.cache/huggingface/datasets/json/default-90b8cf6b2bdf123d/0.0.0/cache-927dd9580a2fd17c.arrow
torch.float32 422326272 0.11537932153507864
torch.uint8 3238002688 0.8846206784649213
{'loss': 1.8124, 'learning_rate': 0.0002, 'epoch': 0.13}                                                                                                                                                        
{'loss': 1.5937, 'learning_rate': 0.0002, 'epoch': 0.26}                                                                                                                                                        
  0%|▍                                                                                                                                                                   | 25/10000 [14:21<101:04:08, 36.48s/it]
```

**Comment by jieguangzhou** [comment]
I think you can delete the cached dataset `/home/ubuntu/.cache/huggingface/datasets/json/default-90b8cf6b2bdf123d/0.0.0`.

**Comment by zhangqibupt** [comment]
I appreciate your suggestion, and I did try that initially. However, it didn't yield the desired results for me. Thank you for reminding me to delve deeper into the issue. I will continue investigating to see if I can identify the root cause. If you have any further ideas or suggestions, please don't hesitate to let me know.

**Comment by ehartford** [comment]
is this tested?  does it create good results?  Can I use it?

**Comment by Compratrex** [comment]
Pretty new to this topic, but it worked for me

On RTX 4070 ti 12G:

100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10000/10000 [1:50:13<00:00, 1.51it/s]
Saving PEFT checkpoint...
***** train metrics *****
  epoch = 48.78
  train_loss = 0.1215
  train_runtime = 1:50:13.32
  train_samples_per_second = 6.048
  train_steps_per_second = 1.512
  
TEST:
python3 -m fastchat.serve.cli --model-path merged_model --device cpu
Loading checkpoint shards: 100%|████████████████████████████████████████████████████████████████████████████| 2/2 [00:10<00:00,  5.24s/it]
Human: Are you trained with IBM data?
Assistant: No, I am trained by researchers from Large Model Systems Organization (LMSYS).
Human: ^Cexit...

Data example: dummy_conversations.json

Will test later on other data.
